app/controllers/users_controller.rb

```diff
@@ -1993,7 +1995,9 @@ def user_menu_messages
     end
 
     if read_notifications.present?
-      Notification.populate_acting_user(read_notifications) if SiteSetting.show_user_menu_avatars
+      if SiteSetting.show_user_menu_avatars || !SiteSetting.prioritize_username_in_ux
+        Notification.populate_acting_user(read_notifications)
+      end
```

Comment:
I didn't see the previous PR but I wonder if we can confine all these logic in the `NotificationSerializer` instead since deciding whether to include an attribute or not is the serializers responsibility. This keeps the conditional logic in a single spot and ensures that all notifications are serialized the same way as long as the conditions are made instead of notifications serialized based on they way the controller has loaded the notifications.

Reply:
@tgxworld We are reviving this PR it seems.

I agree with you but the issue is this cannot be on the notificartion serializer itself, because it would cause an N+1.. the whole reason for `populate_acting_user` is to make 1 user query and map them back to each notification.. https://github.com/discourse/discourse/blob/2ffcb94c8b6d26a2ddc083a6df1b0b174978ee6f/app/models/notification.rb#L379

What we would need is a wrapping serializer AROUND notification like `TopicView` for the post stream, where we can store temporary shiz like this user mapping. We use this pattern in `TopicView` like this: 
https://github.com/discourse/discourse/blob/2ffcb94c8b6d26a2ddc083a6df1b0b174978ee6f/app/serializers/post_serializer.rb#L254-L256